libraries/AP_Common/MultiHeap.cpp

```diff
@@ -112,12 +158,11 @@ void *MultiHeap::allocate(uint32_t size)
  */
 void MultiHeap::deallocate(void *ptr)
 {
-    if (!available()) {
+    if (!available() || ptr == nullptr) {
         return;
     }
-    // NOTE! this relies on either there being a single heap or heap_realloc()
-    // not using the first argument when size is zero.
-    hal.util->heap_realloc(heaps[0], ptr, 0, 0);
+    struct alloc_header *h = ((struct alloc_header *)ptr)-1U;
+    hal.util->heap_realloc(h->heap, h, 0, 0);
```

Comment:
Now that we only use `Util::heap_realloc` as `malloc` or `free`, should it be split into two functions? That would simplify its implementation.

Reply:
yep, i've split it into 2 fns